x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -893,33 +973,91 @@ public void findActiveTokensForRealm(String realmName, ActionListener<Collection
                 .request();
+            securityIndex.checkIndexVersionThenExecute(listener::onFailure,
+                () -> ScrollHelper.fetchAllByEntity(client, request, listener, (SearchHit hit) -> filterAndParseHit(hit, filter)));
+        }
+    }
 
+    public void findActiveTokensForUser(String username, ActionListener<Collection<Tuple<UserToken, String>>> listener) {
+        ensureEnabled();
+
+        final SecurityIndexManager frozenSecurityIndex = securityIndex.freeze();
+        if (Strings.isNullOrEmpty(username)) {
+            listener.onFailure(new IllegalArgumentException("username is required"));
+        } else if (frozenSecurityIndex.indexExists() == false) {
+            listener.onResponse(Collections.emptyList());
+        } else if (frozenSecurityIndex.isAvailable() == false) {
+            listener.onFailure(frozenSecurityIndex.getUnavailableReason());
+        } else {
+            final Instant now = clock.instant();
+            final BoolQueryBuilder boolQuery = QueryBuilders.boolQuery()
+                .filter(QueryBuilders.termQuery("doc_type", "token"))
+                .filter(QueryBuilders.boolQuery()
+                    .should(QueryBuilders.boolQuery()
+                        .must(QueryBuilders.termQuery("access_token.invalidated", false))
+                        .must(QueryBuilders.rangeQuery("access_token.user_token.expiration_time").gte(now.toEpochMilli()))
+                    )
+                    .should(QueryBuilders.termQuery("refresh_token.invalidated", false))
+                );
+
+            final SearchRequest request = client.prepareSearch(SecurityIndexManager.SECURITY_INDEX_NAME)
+                .setScroll(DEFAULT_KEEPALIVE_SETTING.get(settings))
+                .setQuery(boolQuery)
+                .setVersion(false)
+                .setSize(1000)
+                .setFetchSource(true)
+                .request();
             securityIndex.checkIndexVersionThenExecute(listener::onFailure,
-                () -> ScrollHelper.fetchAllByEntity(client, request, listener, this::parseHit));
+                () -> ScrollHelper.fetchAllByEntity(client, request, listener,
+                    (SearchHit hit) -> filterAndParseHit(hit, isOfUser(username))));
         }
     }
 
-    private Tuple<UserToken, String> parseHit(SearchHit hit) {
+    private static Predicate<Map<String, Object>> isOfUser(String username) {
+        return source -> {
+            String auth = (String) source.get("authentication");
+            Integer version = (Integer) source.get("version");
+            Version authVersion = Version.fromId(version);
+            try (StreamInput in = StreamInput.wrap(Base64.getDecoder().decode(auth))) {
+                in.setVersion(authVersion);
+                Authentication authentication = new Authentication(in);
+                return authentication.getUser().principal().equals(username);
+            } catch (IOException e) {
+                return false;
```

Comment:
maybe throw an UncheckedIOException? At the very least, we should log this at some level otherwise we lose the exception and it can make debugging harder.

Reply:
@jkakavas this still needs to be addressed. Let's rethrow as a UncheckedIOException; swallowing and hiding exceptions is bad

Reply:
Thanks for persisting on this, I was sure I had addressed it the first time and thus marked it as resolved